File: ios/LivtetJigsaw/Sources/LivtetJigsaw/Model/Sequence.swift
public typealias InsertsGenerator = (Int, Int) -> Insert

public final class InsertSequence {
    private let generator: InsertsGenerator
    private var index: Int = 0
    private var previous: Insert?

    public init(_ generator: @escaping InsertsGenerator) {
        self.generator = generator
    }

    public func next() -> Insert {
        let value = generator(index, index)
        previous = value
        index += 1
        return value
    }

    public func previousComplement() -> Insert {
        return previous?.complement ?? .none
    }

    public func current(_ totalCount: Int) -> Insert {
        if index == totalCount {
            return .none
        }
        return generator(index, index)
    }

    public static let fixed: InsertsGenerator = { _, _ in .tab }

    public static let twoAndTwo: InsertsGenerator = { index, _ in
        (index % 4) < 2 ? .tab : .slot
    }
}

File: android/jigsaw/src/main/kotlin/net/olamaelcu/livtet/jigsaw/model/Sequence.kt
package net.olamaelcu.livtet.jigsaw.model

typealias InsertsGenerator = (Int) -> Insert

val fixedGenerate: InsertsGenerator = { Insert.Tab }
val flipflopGenerate: InsertsGenerator = { n -> if (n % 2 == 0) Insert.Tab else Insert.Slot }
val twoAndTwoGenerate: InsertsGenerator = { n -> if (n % 4 < 2) Insert.Tab else Insert.Slot }
val randomGenerate: InsertsGenerator = { if (Math.random() < 0.5) Insert.Tab else Insert.Slot }

class InsertSequence(private val generator: InsertsGenerator) {
    private var n: Int = 0
    private var previous: Insert = Insert.None
    private var current: Insert = Insert.None

    fun previousComplement(): Insert = previous.complement()
    fun current(max: Int): Insert = if (n == max) Insert.None else current
    fun next(): Insert { previous = current; current = generator(n++); return current }
}
